### ISSUE TITLE
review service specs for 8-bit compatiblity

### ISSUE THREAD

**Issue by mmoskal** [microsoft/jacdac#273]
This issue tracks status of (possible) implementation of Jacdac services on a low-end 8-bit microcontrollers. A typical example is $0.05 PADAUK with 128 bytes of RAM and 2,000 instructions of one time programmable memory.

Services that are determined to be compliant right now are marked with `tags: 8bit` in the service definition files.

Overall, about 1/3 of services are 8-bit-ready, 1/3 can likely be fixed to be, and 1/3 cannot be fixed.

## Limitations of 8bit services

* no multiplication or division
* max 8 byte payload
* can't buffer lots of data

## Excluded services

The following services are determined to either need to process too much data at once, or are used in situations where the cost savings from the MCU make no sense: arcadesound, barcodereader, bitradio, bootloader, characterscreen, dmx, indexedscreen, iothub, ledmatrix, ledpixel, logger, microphone, midioutput, modelrunner, power, rng, rolemanager, sensoraggregator, settings, soundplayer, soundspectrum, speechsynthesis, tcp, wifi.

Additionally, the various motion sensors require generation of events, and the payloads are likely too large.

* accelerometer - maybe: likely too complex (events)
* gyroscope - maybe: SI units (and has accel problems since integrated)
* magnetomer - maybe: 12 byte payload
* compass - maybe: problem with 3D->degree mapping

## SI units

The following services use SI units, and probably need to be converted to some sort of formula computing the SI unit from the raw readings.

* [ ] barometer
* [ ] distance
* [ ] eco2
* [ ] heartrate
* [ ] humidity
* [ ] illuminance
* [ ] pulseoximeter
* [ ] raingauge
* [ ] servo
* [ ] thermocouple
* [ ] thermometer
* [ ] tvoc
* [ ] uvindex
* [ ] weightscale
* [ ] winddirection
* [ ] windspeed

## Other issues

* [ ] led - maybe: can't do multiple animation steps
* [ ] arcadegamepad - fix: unlimited packet length
* [ ] matrixkeypad - fix: unlimited data size (N buttons)
* [ ] control - fix: flood_ping, status_light
* [ ] prototest - fix: limit packed size to 8?
* [ ] multitouch - unclear
* [ ] vibration - fix: needs size limit

**Comment by mmoskal** [comment]
* SHTC3 - hum/temp - mult+offset OK - $0.88
* analog gas sensors start at around $1 and require different compensation depending on temp

Expensive sensors ($5+) - here it probably doesn't matter, as you will likely just do M0 for flexibility (eg f/w upgrade):

* CCS811 - has eco2, TVOC, temp, humidity in SI units already
* BME280, BME680, BMP390L - all require a very complex compensation

**Comment by mmoskal** [comment]
OK, I'm starting to question the wisdom of "no multiplication" - one can do full software 16x16 bit multiply (with 16 bit result) in under 200 cycles. CRC of 20 byte packet (4 byte payload) is ~500 cycles. So in fact, there should be no problem doing the multiplication on Padauk for most purposes (for 100Hz and slower sampling).

This may limit our options if we eventually go for real ASICs with Jacdac, but I suspect the easiest way to make these would be a built-in Padauk-class MCU - people sometimes do that with M0 on a motion sensors to do data processing. There are good chances I'm wrong on that one though...

**Comment by pelikhan** [comment]
Keeping the protocol simpler is a plus for sure.

________________________________

**Comment by jamesadevine** [comment]
With multiplication, I think the question will be where you slot it in in the processing pipeline. Would you compute it on each sensor read? Would you compute it in the time between the start pulse and data? This might be tricky if this is performed in addition to CRC computation. Maybe we should move the CRC at the end of the frame so that smaller processors can compute crc in spare cycles during transmission. 

There's also the question around further packet formatting after the multiplication has taken place.

**Comment by pelikhan** [comment]
> Maybe we should move the CRC at the end of the frame so that smaller processors can compute crc in spare cycles during transmission.

This is another breaking change to the protocol format.

**Comment by mmoskal** [comment]
I would be inclined to put CRC in the start pulse gap, and leave the multiplication etc to the "application code". I imagine the architecture would be:
* there's an interrupt waiting for low pulse (either timer or pin IRQ); when this happens, we do reception completely in interrupt handler
* there's the main processing loop ("app code") where we talk I2C, do multiplications and what's not
* when app code is ready to send, it constructs a packet in memory, and sets "send" flag
* if enough time has passed since reception we start send (in app code, with IRQ disabled)

In other words, the app code doesn't have to fit in some tight timelines. Note that if we're talking I2C with something, this is going to take on the order of 1000 cycles, so it doesn't seem the 200 for multiplication is going to make much of a difference.

**Comment by tballmsft** [comment]
Is 8-bit really the best choice?  MCUs are still riding Moore's Law. It would be a pity if we kneecapped Jacdac and rued our decision in a few years. Does going to 16-bit MCU buy us anything, so to speak?

**Comment by mmoskal** [comment]
I think the space for 16-bit is shrinking real fast. Anyway, I think we're good to support Padauk right now, the question is whether we should consider future implementations of Jacdac in silicon (ASICs). I think most of these have a Padauk-like core in them anyway, so we may be good here as well...

**Comment by tballmsft** [comment]
OK. As long as we don't compromise a service specification to fit into 8-bits.

**Comment by mmoskal** [comment]
Several services were implemented in https://github.com/microsoft/jacdac-padauk/tree/main/services

This includes illuminance which uses SI units, so I think our worries were misplaced there. I have tagged all simple SI unit services with 8bit in #370 

The one remaining one is `servo`; need to figure out if padauk can do the math

**Comment by pelikhan** [comment]
@mmoskal looks like this workitem is done?
